modules/mtl/arrow-mtl-data/src/test/kotlin/arrow/mtl/StateTTests.kt

```diff
@@ -1,64 +1,99 @@
 import arrow.test.UnitSpec
+import arrow.test.generators.GenK
+import arrow.test.generators.tuple2
 import arrow.test.laws.AsyncLaws
 import arrow.test.laws.MonadCombineLaws
 import arrow.test.laws.MonadStateLaws
 import arrow.test.laws.SemigroupKLaws
 import arrow.typeclasses.Eq
+import arrow.typeclasses.EqK
+import arrow.typeclasses.Monad
+import io.kotlintest.properties.Gen
 
 class StateTTests : UnitSpec() {
 
   val M: StateTMonadState<ForTry, Int> = StateT.monadState(Try.monad())
 
-  val EQ: Eq<StateTOf<ForTry, Int, Int>> = Eq { a, b ->
-    a.runM(Try.monad(), 1) == b.runM(Try.monad(), 1)
-  }
+  fun <F, S> eqk(EQKF: EqK<F>, EQS: Eq<S>, M: Monad<F>, s: S) = object : EqK<StateTPartialOf<F, S>> {
+    override fun <A> Kind<StateTPartialOf<F, S>, A>.eqK(other: Kind<StateTPartialOf<F, S>, A>, EQ: Eq<A>): Boolean =
+      (this.fix() to other.fix()).let {
+        val ls = it.first.runM(M, s)
+        val rs = it.second.runM(M, s)
 
-  val EQ_UNIT: Eq<StateTOf<ForTry, Int, Unit>> = Eq { a, b ->
-    a.runM(Try.monad(), 1) == b.runM(Try.monad(), 1)
+        EQKF.liftEq(Tuple2.eq(EQS, EQ)).run {
+          ls.eqv(rs)
+        }
+      }
   }
 
-  val EQ_LIST: Eq<Kind<StateTPartialOf<ForListK, Int>, Int>> = Eq { a, b ->
-    a.runM(ListK.monad(), 1) == b.runM(ListK.monad(), 1)
+  val listkStateEQK = eqk(ListK.eqK(), Int.eq(), ListK.monad(), 1)
+
+  val ioStateEQK = eqk(IO.eqK(), Int.eq(), IO.monad(), 1)
+
+  val tryEQK = object : EqK<ForTry> {
+    override fun <A> Kind<ForTry, A>.eqK(other: Kind<ForTry, A>, EQ: Eq<A>): Boolean =
+      (this.fix() to other.fix()).let {
+        Option.eq(EQ).run {
+          it.first.toOption().eqv(it.second.toOption())
+        }
+      }
   }
 
-  private fun <A> IOEQ(): Eq<StateTOf<ForIO, Int, A>> = Eq { a, b ->
-    a.runM(IO.monad(), 1).attempt().unsafeRunSync() == b.runM(IO.monad(), 1).attempt().unsafeRunSync()
+  val tryStateEqK: EqK<Kind<Kind<ForStateT, ForTry>, Int>> = eqk(tryEQK, Int.eq(), Try.monad(), 1)
+
+  fun <F, S> genk(genkF: GenK<F>, genS: Gen<S>) = object : GenK<StateTPartialOf<F, S>> {
+    override fun <A> genK(gen: Gen<A>): Gen<Kind<StateTPartialOf<F, S>, A>> =
+      genkF.genK(genkF.genK(Gen.tuple2(genS, gen)).map { state ->
+        val stateTFun: StateTFun<F, S, A> = { _: S -> state }
+        stateTFun
+      }).map {
+        StateT(it)
+      }
   }
 
   init {
     testLaws(
-      MonadStateLaws.laws(M, EQ, EQ_UNIT),
-      AsyncLaws.laws<StateTPartialOf<ForIO, Int>>(StateT.async(IO.async()), IOEQ(), IOEQ()),
+      MonadStateLaws.laws(M, tryStateEqK),
+      AsyncLaws.laws<StateTPartialOf<ForIO, Int>>(StateT.async(IO.async()), ioStateEQK),
+
       SemigroupKLaws.laws(
         StateT.semigroupK<ForListK, Int>(ListK.monad(), ListK.semigroupK()),
-        StateT.applicative<ForListK, Int>(ListK.monad()),
-        EQ_LIST),
+        Gen.int().map { StateT.applicative<ForListK, Int>(ListK.monad()).just(it) } as Gen<Kind<StateTPartialOf<ForListK, Int>, Int>>,
```

Comment:
@1Jajen1 making progress. of the laws i modified in the PR this should be the last spot where `Gen` is required from within tests. I could make all other laws functions internal.

Reply:
I cloned your project to look at the slow/failing tests. So far this is coming along nicely, I am surprised there are not more test failures :) Seems like arrow is quite well written :rofl: